analytics/tests/test_activity_views.py

```diff
@@ -218,3 +220,148 @@ def test_get_remote_server_guest_and_non_guest_count(self) -> None:
         remote_activity_counts = get_remote_customer_user_count(server_logs[server_id])
         self.assertEqual(remote_activity_counts.non_guest_user_count, 73)
         self.assertEqual(remote_activity_counts.guest_user_count, 16)
+
+    def test_remote_activity_with_robust_data(self) -> None:
+        def add_plan(customer: Customer, tier: int, fixed_price: bool = False) -> None:
+            if tier in (
+                CustomerPlan.TIER_SELF_HOSTED_BASE,
+                CustomerPlan.TIER_SELF_HOSTED_LEGACY,
+                CustomerPlan.TIER_SELF_HOSTED_COMMUNITY,
+            ):
+                price_per_license = 0
+            else:
+                price_per_license = 1000
+            if not fixed_price:
+                plan = CustomerPlan.objects.create(
+                    customer=customer,
+                    billing_cycle_anchor=timezone_now(),
+                    billing_schedule=CustomerPlan.BILLING_SCHEDULE_ANNUAL,
+                    tier=tier,
+                    price_per_license=price_per_license,
+                    next_invoice_date=add_months(timezone_now(), 12),
+                )
+            else:
+                plan = CustomerPlan.objects.create(
+                    customer=customer,
+                    billing_cycle_anchor=timezone_now(),
+                    billing_schedule=CustomerPlan.BILLING_SCHEDULE_ANNUAL,
+                    tier=tier,
+                    fixed_price=10000,
+                    next_invoice_date=add_months(timezone_now(), 12),
+                )
+            LicenseLedger.objects.create(
+                licenses=10,
+                licenses_at_next_renewal=10,
+                event_time=timezone_now(),
+                is_renewal=True,
+                plan=plan,
+            )
```

Comment:
This is OK but I usually find it nicer to write tests that create objects via planning parameters to pass to `process_initial_upgrade` rather than manually creating objects -- mainly because that sort of end-to-end testing requires less maintenance. (If we change how realms get upgraded, that's more likely to require manual intervention in this test).

It may not be worth changing though -- the maintenance time savings may be less than the time required to do so -- but just wanted to note that general principle.

Reply:
(And in particular `create_plan_for_customer` is using the same pattern as this test)

Reply:
Yeah, we'd have to mock a whole lot of stripe calls to do it that way. I was thinking of these tests as mostly checking our logic around our database table queries for these views. And the existing test wasn't really showing the number of queries when there were multiple remote servers/realms with active plans.

But I can definitely go back and work on revising for that format.